### PR TITLE
[ISSUE 34] ADR: Recording architectural decisions

### PR DESCRIPTION
**Pull request by widal001**
### Description

This PR introduces a new ADR (architecture decision record) template and directory, and it proposes an ADR for selecting a wiki platform based on the objectives outlined in the communications platform milestone.

Closes: #34 

### Changes Proposed

- Creates an `documentation/adr/` directory which will store architecture decision records related to the project
- Proposes `adr/0001-recording-architecture-decisions.md` which describes a processing for using ADRs to record architecture decisions throughout the project

### Instructions to Review

1. Review the `documentation/adr/README.md` to understand the purpose and the objective of ADRs
2. Review the `adr/template.md` and offer suggestions for changes to this template
3. Review the `adr/0001-recording-architecture-decisions.md` and recommend changes to the ADR process

### Notes

The Wiki ADR has been moved to another branch

### CODE REVIEW

File: documentation/adr/template.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  - Good, because {argument a}
+  - Good, because {argument b}
+  - ...
+- **Cons**
+  - Bad, because {argument c}
+  - ...
+
+## Links <!-- OPTIONAL -->
+
+- [{Link type}](link to adr) <!-- example: Refined by [xxx](yyyymmdd-xxx.md) -->
```
Comment (lucasmbrown-usds) [inline documentation/adr/template.md:67]: nit: i think this example is copy/pasta from a previous line?

Comment (widal001) [reply documentation/adr/template.md:67]: Yeah good call! I meant this more for external resources that might be related to the ADR. I'll update this line to be:
```suggestion
- [{Link name}](link to external resource)
```

File: documentation/adr/0001-recording-architecture-decisions.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+- **Documented:** There should be a written record describing which decisions are made
+- **Accessible:** Decision records should be centrally located, searchable, and accessible to non-technical audiences
+- **Portable:** The decision records should be able to travel with the code when it's forked, cloned, etc.
+- **Maintainable:** The system we adopt should be easy to maintain or amend as needed.
+
+## Options Considered
+
+- ADRs in the repo `docs/` folder
+- Comments and docstrings in the code itself
+- Articles in the project Wiki tab
```
Comment (lucasmbrown-usds) [inline documentation/adr/0001-recording-architecture-decisions.md:27]: Should we also consider using the overall wiki we choose for the comms platform?

File: documentation/adr/0002-wiki-platform.md [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+## Context and Problem Statement
+
+The [communications platform milestone](milestone) identifies a series of platforms through which the Grants API project needs to engage both internal and external stakeholders. One of these platforms is a wiki for storing notes, documents, and other content about the project. Ideally we would select a platform that balances ease of use and flexibility with the cost of implementing and maintaining the wiki.
+
+The goal of this ADR is to evaluate a series of potential wiki platforms and determine which one best fits the needs and objectives of this project based on the decision criteria outlined below.
+
+## Decision Drivers <!-- RECOMMENDED -->
+
+#### Must Have
```
Comment (lucasmbrown-usds) [inline documentation/adr/0002-wiki-platform.md:17]: Let's add AuthZ model per page: the ability to set some pages as viewable to the public, some as private to a group of users, etc. As well as the ability to control who can edit which content. We might want some pages that are publicly viewable but locked to editing by a small group.

Comment (lucasmbrown-usds) [reply documentation/adr/0002-wiki-platform.md:17]: And let's add accessibility (for visually impaired and other users).

Also, available on restricted networks at HHS and, as a stretch goal, the White House (EOP).

File: documentation/adr/0002-wiki-platform.md [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+- **Usability:** Non-technical users should be able to access and create content with minimal training or guidance.
+- **Content Review:** Collaborators should be able to review and edit draft content before those changes are published.
+- **Version History:** Editors should be able to see and restore previous versions of a given page.
+- **Multi-Media:** The platform should support multiple types of media (e.g. videos, images, file uploads, tables, diagrams) with minimal configuration.
+- **Internationalization (i18n):** The platform should provide support for displaying content in multiple languages.
+- **Web Analytics:** The platform should provide support for tracking site usage and other web analytics.
+- **Onboarding Costs:** Onboarding new members to the platform should be relatively inexpensive, both in terms of staff time/resources and direct costs (e.g. licensing fees).
+- **Maintenance Costs:** It should not be prohibitively expensive to maintain the wiki, both in terms of staff time/resources and direct costs (e.g. hosting fees).
+
+#### Nice to Have
```
Comment (lucasmbrown-usds) [inline documentation/adr/0002-wiki-platform.md:28]: do you want anything about machine readability or structured data here as a nice-to-have? it seems related to 'data access' but not quite the same thing.

File: documentation/adr/0002-wiki-platform.md [hunk 5]
```diff
@@ -0,0 +1,294 @@
+# Communications Tooling: Wiki Platform
```
Comment (lucasmbrown-usds) [inline documentation/adr/0002-wiki-platform.md:1]: The more I think about it, the more I think we move this into a separate PR. Or just merge it incomplete and open a new PR to actually make the decision.

File: documentation/adr/0002-wiki-platform.md [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+## Pros and Cons of the Options <!-- OPTIONAL -->
+
+### Confluence
+
+[Confluence](confluence) is a Software as a Service (SaaS) documentation and collaboration tool offered by Atlassian that organizes content into "spaces" and offers a series of templates and components that can be used to create custom documentation for internal and external stakeholders.
+
+#### Details
+
+- **Hosting:** SaaS
+- **Pricing:** [$5.75 (standard) or $11 (premium) per user per month](confluence-pricing)
```
Comment (lucasmbrown-usds) [inline documentation/adr/0002-wiki-platform.md:67]: Is this the cost even just to view content, or for edit users only?

File: documentation/adr/0002-wiki-platform.md [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+- Data is controlled by Confluence, only accessible via API
+
+### Notion
+
+[Notion](notion) is a Software as a Service (SaaS) documentation and collaboration tool that also allows users to add structured and semi-structured content to pages. Because Notion offers a fully-featured API for reading and managing content, it also has a robust set of community integrations that extend Notion's core functionality.
+
+#### Details
+
+- **Hosting:** SaaS
+- **Pricing:** [$8 (pro) or $15(business) per user per month](notion-pricing)
```
Comment (lucasmbrown-usds) [inline documentation/adr/0002-wiki-platform.md:102]: same pricing question

File: documentation/adr/0002-wiki-platform.md [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+- Exposes wiki content via an API
+- Robust plugin and add-on community
+
+#### Cons
+
+- Slightly more complicated interface than Confluence
+- Pages published to the web aren't organized as clearly as public pages in GitBook
+- No content review process, all changes and comments are published automatically
+- No direct support for internationalization and localization
+- Page history is limited to 30 (pro) or 90 (business) days
```
Comment (lucasmbrown-usds) [inline documentation/adr/0002-wiki-platform.md:126]: oof

**Review by lucasmbrown-usds — CHANGES_REQUESTED** [review]
Overall, this looks very good. I think my comments are mostly just on the individual decisions themselves, not on the templates. I should be a quick 👍 soon!

File: documentation/adr/0001-recording-architecture-decisions.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
+# Recording Architecture Decisions
```
[the remark this answers is not in the document]
Comment (widal001) [reply documentation/adr/0001-recording-architecture-decisions.md:1]: Yeah I like that! That would make it easier to find the template too. What are your thoughts on naming it `adr/decisions/`?

File: documentation/adr/0001-recording-architecture-decisions.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+_What is the best way to document key architectural decisions made within the project so that future contributors can understand the justification for those decisions?_
+
+## Decision Drivers <!-- RECOMMENDED -->
+
+- **Explicit:** Important decisions about the project architecture should be clear and unambiguous
+- **Documented:** There should be a written record describing which decisions are made
+- **Accessible:** Decision records should be centrally located, searchable, and accessible to non-technical audiences
+- **Portable:** The decision records should be able to travel with the code when it's forked, cloned, etc.
+- **Maintainable:** The system we adopt should be easy to maintain or amend as needed.
```
[the remark this answers is not in the document]
Comment (widal001) [reply documentation/adr/0001-recording-architecture-decisions.md:21]: Yeah this is a good question! Hmmm ultimately I was thinking these could be moved into the wiki which relates to your comment below:

> I might be tempted to vote for an alternative we haven't considered: putting the ADRs in a separate wiki platform (whatever we end up using for our wiki) rather than in the repo, but I could be persuaded either way. Let's discuss verbally sometime.

That would make them easier to edit for non-technical users. 

What's worth noting is that GitBook could allow us to achieve both of these goals with their [GitHub integration](https://docs.gitbook.com/product-tour/git-sync/enabling-github-sync) which syncs the content between a GitHub repo and GitBook itself and allows you to [specify a root path for the docs](https://docs.gitbook.com/product-tour/git-sync/content-configuration#root)

Reading through this list again, though, I do think there's some inherent tension between the `centrally located` condition in the Accessible criterion and `travel with the code when it's forked, cloned, etc.` in the Portable criterion. It could be helpful to think through which of these is the more important goal. and if there's some other way to link ADRs to the codebase that they are relevant to if we decide not to have a mono-repo.

File: documentation/adr/0001-recording-architecture-decisions.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+### Positive Consequences <!-- OPTIONAL -->
+
+- ADRs travel with the repo when it's cloned, forked, etc.
+- ADRs can be incorporated in the Issue and PR workflow
+- Changes to ADRs will be listed in the project's commit history
+- ADRs written in Markdown are accessible to non-technical audiences
+
+### Negative Consequences <!-- OPTIONAL -->
+
+- Certain urgent decisions may take longer to finalize if they need to be documented and agreed upon, slowing down
```
[the remark this answers is not in the document]
Comment (widal001) [reply documentation/adr/0001-recording-architecture-decisions.md:42]: Thanks for catching this! I'll fix it in the next push.

File: documentation/adr/0001-recording-architecture-decisions.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+## Decision Outcome <!-- REQUIRED -->
+
+Decisions will be documented using Architecture Decision Records (ADRs) as described by [Joel Parker Henderson](joel) and the [ADR GitHub organization](https://adr.github.io/). The template for this project's ADRs will be adapted from the [MADR template](adrs).
+
+### Positive Consequences <!-- OPTIONAL -->
+
+- ADRs travel with the repo when it's cloned, forked, etc.
+- ADRs can be incorporated in the Issue and PR workflow
+- Changes to ADRs will be listed in the project's commit history
+- ADRs written in Markdown are accessible to non-technical audiences
```
[the remark this answers is not in the document]
Comment (widal001) [reply documentation/adr/0001-recording-architecture-decisions.md:38]: Yeah this is a fair point. I was considering it accessible from reading the resulting ADRs (especially compared to inline code comments or docstrings). But in order to make it more accessible to write ADRs we'd probably need some kind of GUI wrapper, which again relates to your suggestion below:

> I might be tempted to vote for an alternative we haven't considered: putting the ADRs in a separate wiki platform (whatever we end up using for our wiki) rather than in the repo, but I could be persuaded either way. Let's discuss verbally sometime.

File: documentation/adr/0001-recording-architecture-decisions.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+  - Docstrings travel with the code when it's forked, cloned, etc.
+  - Changes to docstrings and comments are part of the commit history
+- **Cons**
+  - Not very accessible to non-technical audiences
+  - Not centralized and difficult to search for
+  - Hard to see the sequential evolution of decisions
+
+### Project Wiki
+
+Documenting decisions in the Wiki tab on the project repository in GitHub.
```
[the remark this answers is not in the document]
Comment (widal001) [reply documentation/adr/0001-recording-architecture-decisions.md:61]: Yeah I think the ideal solution here would be to have a workflow where users can view/create/edit ADRs in a wiki-style interface that also syncs the content with a GitHub repo. 

That may be simple with a tool like GitBook or require some creativity if we decide to go with Confluence or Notion. But would love to talk through this stuff on Tuesday when we meet with @acouch

File: documentation/adr/0002-wiki-platform.md (hunk 4 above)
Comment (widal001) [reply documentation/adr/0002-wiki-platform.md:28]: Yeah that's a really good suggestion. Machine readable and/or structured content feels like it's own goal for sure and definitely a differentiator across wiki platforms

File: documentation/adr/0002-wiki-platform.md (hunk 5 above)
Comment (widal001) [reply documentation/adr/0002-wiki-platform.md:1]: Agreed, I wasn't planning to really make any more edits to this ADR in this branch. Do you think it would be better to leave as is, or just copy the data into a different branch and delete this file?

File: documentation/adr/0002-wiki-platform.md (hunk 6 above)
Comment (widal001) [reply documentation/adr/0002-wiki-platform.md:67]: All of the pricing is for named users, what I was imagining would be HHS staff and contractors whom we want to have direct edit and content moderation access. 

The ability for members of the public to view, comment on, and suggest edits is covered by the external contributions section and in the Pros and Cons list, which often references features like publishing spaces or pages to the web. But I can definitely make that distinction a bit clearer!

File: documentation/adr/0002-wiki-platform.md (hunk 8 above)
Comment (widal001) [reply documentation/adr/0002-wiki-platform.md:126]: Yeah I was impressed by a lot of Notion features, especially their API, but the cons are pretty bad 😞

File: documentation/adr/0002-wiki-platform.md (hunk 7 above)
Comment (widal001) [reply documentation/adr/0002-wiki-platform.md:102]: Same note as above. All prices are for named users with edit access to the wiki.

File: documentation/adr/0001-recording-architecture-decisions.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (lucasmbrown-usds) [reply documentation/adr/0001-recording-architecture-decisions.md:1]: Good flags @acouch on having too many decisions! Maybe we start with a single folder for now, but plan to revisit it shortly as our decisions pile up and we can then add a little more information architecture structure? 

So just to be clear on our naming conventions, I believe we're suggesting we go with `<repo-root>/documentation/decisions` for the folder for everything related to ADRs. What's the name of the directory for the individual decisions? should that be `<repo-root>/documentation/decisions/individual_decisions` to mimic `milestones/individual_milestones`, or should we come up with a better naming convention for both?

File: documentation/adr/0001-recording-architecture-decisions.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (lucasmbrown-usds) [reply documentation/adr/0001-recording-architecture-decisions.md:21]: Yeah it does seem like those are in tension. I would personally lean towards `centrally located` over `travel with the code...`, but I could definitely be persuaded in another direction!

File: documentation/adr/0002-wiki-platform.md (hunk 5 above)
Comment (lucasmbrown-usds) [reply documentation/adr/0002-wiki-platform.md:1]: Either is fine for me! It seems simpler to leave as is and then open a new branch to update the existing file, but either works.

File: documentation/adr/0001-recording-architecture-decisions.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (lucasmbrown-usds) [reply documentation/adr/0001-recording-architecture-decisions.md:21]: Can you also add this part as a decision driver for this: 

> Should also be part of the public record, in the open, etc.

File: documentation/adr/0002-wiki-platform.md (hunk 3 above)
Comment (lucasmbrown-usds) [reply documentation/adr/0002-wiki-platform.md:17]: If we open a new PR just for `documentation/adr/0002-wiki-platform.md`, let's make sure it incorporates these comments?

File: documentation/adr/0002-wiki-platform.md (hunk 6 above)
Comment (lucasmbrown-usds) [reply documentation/adr/0002-wiki-platform.md:67]: Making that distinction clearer, in this PR or the next one, would be great! Thx!

File: documentation/adr/README.md [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+4. **Risk Management:** Documenting key decisions also allows project stakeholders to assess the potential impact of architectural changes, identify dependencies, and evaluate the consequences of modifying specific components. By having a clear understanding of the trade-offs associated with each option, teams can make more informed decisions and mitigate potential risks associated with changes or enhancements.
+5. **Improved Maintainability:** Software projects evolve over time due to changing requirements, new features, or technological advancements. Documentation ensures that future developers can understand the architecture's design principles, patterns, and constraints. It enables future maintainers to modify or extend the codebase in a consistent and coherent manner, reducing the risk of introducing unintended side effects or architectural drift.
```
Comment (lucasmbrown-usds) [inline documentation/adr/README.md:14]: Maybe "Maintainability and Adaptability."? To indicate that it also helps us be more confident in changing our minds later. 

I remember on CEJST we switched from a mapping framework (whose name I'm currently blanking on) back to MapLibre after playing with the former for a while. By revisiting our decision log, it was very easy to see that given our new information, MapLibre was the best choice.

File: documentation/adr/template.md [hunk 15]
```diff
@@ -0,0 +1,68 @@
+# {short title of solved problem and solution}
+
+- **Status:** {Draft | Proposed | Rejected | Accepted | Deprecated | ... | Superseded By [xxx](yyyymmdd-xxx.md)} <!-- REQUIRED -->
+- **Last Modified:** {YYYY-MM-DD when the decision was last updated} <!-- REQUIRED -->
+- **Related Issue:** [#{issue number}](https://github.com/HHS/grants-api/issues) <!-- RECOMMENDED -->
+- **Deciders:** {list everyone involved in the decision} <!-- OPTIONAL -->
```
Comment (lucasmbrown-usds) [inline documentation/adr/template.md:6]: should this be required?

**Review by lucasmbrown-usds — APPROVED** [review]
Left a few comments and suggestions throughout, but overall looking very good! 

Would love for @acouch or his delegate to sign off before merging, too.

File: documentation/adr/0002-wiki-platform.md (hunk 4 above)
Comment (widal001) [reply documentation/adr/0002-wiki-platform.md]: This file was moved into #30

File: documentation/adr/0002-wiki-platform.md (hunk 5 above)
Comment (widal001) [reply documentation/adr/0002-wiki-platform.md]: This file was moved into #30

File: documentation/adr/0002-wiki-platform.md (hunk 6 above)
Comment (widal001) [reply documentation/adr/0002-wiki-platform.md]: File moved and comment addressed in #30

File: documentation/adr/template.md (hunk 15 above)
Comment (widal001) [reply documentation/adr/template.md]: Good call! I updated this in [this commit](https://github.com/HHS/grants-api/pull/15/commits/0e4207784335a5e4e8a6da53ce7fa979b9ea075e)

File: documentation/adr/README.md (hunk 14 above)
Comment (widal001) [reply documentation/adr/README.md]: Yeah I remember that switch! It was really helpful to see the history of that decision. 

I changed this to `Improved Maintainability and Adaptability` in [this commit](https://github.com/HHS/grants-api/pull/15/commits/0e4207784335a5e4e8a6da53ce7fa979b9ea075e)

File: documentation/adr/0001-recording-architecture-decisions.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (widal001) [reply documentation/adr/0001-recording-architecture-decisions.md:64]: I tweaked the language of the third option to be a bit more general to any project wiki, not just the GitHub wiki tab, and also added a note that we should revisit our ADR decision when we've onboarded the wiki platform.

The updated language is in [this commit](https://github.com/HHS/grants-api/pull/15/commits/0e4207784335a5e4e8a6da53ce7fa979b9ea075e)

File: documentation/adr/0001-recording-architecture-decisions.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (widal001) [reply documentation/adr/0001-recording-architecture-decisions.md]: I wound up changing this from a pro to a con for the current decision. I'm hoping there's a way we can have our cake and eat it to with the wiki platform choice, but for now having them be stored and edited in GitHub is probably more of a con from readability. 

This change was made in [this commit](https://github.com/HHS/grants-api/pull/15/commits/0e4207784335a5e4e8a6da53ce7fa979b9ea075e)

File: documentation/adr/0001-recording-architecture-decisions.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (widal001) [reply documentation/adr/0001-recording-architecture-decisions.md]: Made this change in [this commit](https://github.com/HHS/grants-api/pull/15/commits/0e4207784335a5e4e8a6da53ce7fa979b9ea075e)

File: documentation/adr/0001-recording-architecture-decisions.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (widal001) [reply documentation/adr/0001-recording-architecture-decisions.md:21]: Added this in [this commit](https://github.com/HHS/grants-api/pull/15/commits/0e4207784335a5e4e8a6da53ce7fa979b9ea075e)

File: documentation/adr/0001-recording-architecture-decisions.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (widal001) [reply documentation/adr/0001-recording-architecture-decisions.md]: @acouch @lucasmbrown-usds I've addressed all other comments on this PR and for this suggestion I restructured the folder like so:

```
documentations/
    adr/
        0001-recording-architecture-decisions.md
    README.md
    template.md
```

Let me know if this works for the time being or you want something different. Agree that we may want to consider splitting them out into sub-folders if the `decisions/adr/` starts to get too big, though I admit that I do like the idea of having all decisions be represented chronologically in a single folder.

File: documentation/decisions/adr/0001-recording-architecture-decisions.md [hunk 16]
```diff
@@ -1,9 +1,9 @@
 # Recording Architecture Decisions
 
 - **Status:** Proposed
-- **Last Modified:** 2023-06-07
-- **Related Issue:** TODO
-- **Deciders:** TODO
+- **Last Modified:** 2023-06-26
+- **Related Issue:** [#34](https://github.com/HHS/grants-api/issues/34)
+- **Deciders:** Lucas Brown, Aaron Couch, Billy Daly
```
Comment (lucasmbrown-usds) [inline documentation/decisions/adr/0001-recording-architecture-decisions.md:6]: :tada:

File: documentation/decisions/adr/0001-recording-architecture-decisions.md [hunk 17]
```diff
@@ -19,29 +19,32 @@ _What is the best way to document key architectural decisions made within the pr
...
 - ADRs in the repo `docs/` folder
 - Comments and docstrings in the code itself
-- Articles in the project Wiki tab
+- Articles in the project's Wiki
 
 ## Decision Outcome <!-- REQUIRED -->
 
 Decisions will be documented using Architecture Decision Records (ADRs) as described by [Joel Parker Henderson](joel) and the [ADR GitHub organization](https://adr.github.io/). The template for this project's ADRs will be adapted from the [MADR template](adrs).
 
+**NOTE:** We may want to revisit this decision to reconsider storing them in the Project Wiki once that wiki is established. Ideally ADRs could be drafted/edited and visible in the wiki but also synced with the repo -- some wiki solutions provide this option.
```
Comment (lucasmbrown-usds) [inline documentation/decisions/adr/0001-recording-architecture-decisions.md:34]: 👍

**Review by lucasmbrown-usds — APPROVED** [review]
Looks great! Nice work!!

File: documentation/adr/0001-recording-architecture-decisions.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (acouch) [reply documentation/adr/0001-recording-architecture-decisions.md]: This looks great to get started.

**Review by acouch — APPROVED** [review]
